<a href="https://colab.research.google.com/github/Victor-Tal/Sorting-Algorithms/blob/main/Sorting_Algortithms.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Titulo del Cuaderno de Trabajo de Colab
<center>
Victor  Hugo Cordero<br>
11 de febrero de 2022<hr>
</center>


El primer programa que haremos sera $\texttt{Insertion-sort}$. Recordemos que la entrada es un conjunto de datos: $a_1, a_2, \ldots, a_n$ almacenado en un arreglo $A$ y la salida es un reordenamiento de los datos, denotado como, $a_1', a_2', \ldots, a_n'$ tale que :

$$ a_1'< a_2' < \ldots < a_n' $$

#Insertion Sort

In [ ]:
def Insertion_Sort(A):
  '''
  Algoritmo Insertion_Sort recibe una lista 
  y la ordena de menor a mayor 
  '''
  for j in range(len(A)):
    k = A[j]
    i = j-1
    while i>=0 and A[i]>k:
      #Para que sea dee menor a mayor, hay que cambiar A[i]<k
      A[i+1] = A[i]
      i-=1
    A[i+1] = k
  return A

In [ ]:
Desordenada = [1231,453,5,145,86,4]
print(Insertion_Sort(Desordenada))

[4, 5, 86, 145, 453, 1231]


#Merge-Sort

In [ ]:
def Merge(Li,Ld):
  L = []
  while len(Li)>0 and len(Ld)>0:
    if Li[0]<Ld[0]:
      L.append(Li[0])
      Li.pop(0)
    else:
      L.append(Ld[0])
      Ld.pop(0)
  while len(Li)>0:
    L.append(Li[0])
    Li.pop(0)
  while len(Ld)>0:
    L.append(Ld[0])
    Ld.pop(0)
  return L

def MergeSort(A):
  if len(A)>1:
    Li = MergeSort(A[0:len(A)//2])
    Ld = MergeSort(A[len(A)//2 : len(A)])
    return Merge(Li,Ld)
  else:
    return A 

In [ ]:
print(MergeSort([6,7,9,13,2,5,19,20]))

[2, 5, 6, 7, 9, 13, 19, 20]


#Burbuja


In [ ]:
def burbuja(A):
  n= len(A)
  for i in range(n):
    for j in range (n-i):
      for k in range(j):
        if A[k]>A[k+1]:
          aux = A[k+1]
          A[k+1]= A[k]
          A[k]= aux
  return A

In [ ]:
B = [9,6,13,2,19,20,5]
print(burbuja(B))


[2, 5, 6, 9, 13, 19, 20]


#Merge Sort Cormen


In [ ]:
inf = float('inf')
def Merge_Cormen(A,p,q,r):
  #p=<q=<r
  n1 = q-p+1 #n1>=1
  n2 = r-q  
  L=[0]*(n1+1)
  R=[0]*(n2+1)

  i,j = 0,0
  for i in range(n1):
    L[i] = A[p+i]
  for j in range(n2):
    R[j] = A[q+j+1]
  L[n1] = inf
  R[n2] = inf
  
  i = 0
  j= 0
  for k in range (p,r+1):
    if L[i]<=R[j]:
      A[k]= L[i]
      i+=1
    else:
      A[k] = R[j]
      j+=1
  return A

def MergeSort_Cormen(A,p,r):
  if p<r:
    q = (p+r)//2

    MergeSort_Cormen(A,p,q)
    MergeSort_Cormen(A,q+1,r)
    return Merge_Cormen(A,p,q,r)


In [ ]:
A= [1,2,65,12,12,312,564,43534,212.53,1,212.54]
print(MergeSort_Cormen(A,1,len(A)-1))

[1, 1, 2, 12, 12, 65, 212.53, 212.54, 312, 564, 43534]


#Comparativa de tiempos Algoritmos


In [ ]:
import random
import time
import copy

N = 2**10
A = N*[0]
for i in range(N):
  A[i] = random.random()

print('Para N= 2^11')
AB = copy.deepcopy(A)
ti = time.time()
burbuja(AB)
t = time.time() - ti
print('Burbuja tardó: \t\t\t', t, ' \tsegundos')

AI = copy.deepcopy(A)
ti = time.time()
Insertion_Sort(AI)
t = time.time() - ti
print('Insertion_Sort tardó: \t\t', t, ' \tsegundos')

AM = copy.deepcopy(A)
ti = time.time()
MergeSort(AM)
t = time.time() - ti
print('Merge Sort tardó: \t\t', t, ' \tsegundos')

AMP = copy.deepcopy(A)
ti= time.time()
AMP.sort()
t = time.time() - ti
print('El metodo sort de python tardó: ', t, ' segundos')



AMC = copy.deepcopy(A)
ti= time.time()
MergeSort_Cormen(AMC,1,len(AMC)-1)
t = time.time() - ti
print('Merge Sort del Cormen tardó: \t', t, ' segundos')


Para N= 2^11
Burbuja tardó: 			 20.698434829711914  	segundos
Insertion_Sort tardó: 		 0.058448076248168945  	segundos
Merge Sort tardó: 		 0.0059795379638671875  	segundos
El metodo sort de python tardó:  0.00018072128295898438  segundos
Merge Sort del Cormen tardó: 	 0.0052776336669921875  segundos


# Para N= 10^2 <hr>

Burbuja tardó: 			 0.019141674041748047  	segundos

Insertion_Sort tardó: 		 0.0005266666412353516  	segundos

Merge Sort tardó: 		 0.0004897117614746094  	segundos

El metodo sort de python tardó:  4.839897155761719e-05  segundos

Merge Sort del Cormen tardó: 	 0.00046563148498535156  segundos

#Para N= 10^3 <hr>

Burbuja tardó: 			 19.015961408615112  	segundos 

Insertion_Sort tardó: 		 0.06224942207336426  	segundos

Merge Sort tardó: 		 0.0058443546295166016  	segundos

El metodo sort de python tardó:  0.00016736984252929688  segundos

Merge Sort del Cormen tardó: 	 0.005988121032714844  segundos



# Para N= 10^4 <hr>
Burbuja tardo: + de 30 minutos

Insertion_Sort tardó: 		 5.855761528015137  	segundos

Merge Sort tardó: 		 0.08006739616394043  	segundos

El metodo sort de python tardó:  0.001619577407836914  segundos

Merge Sort del Cormen tardó: 	 0.06514286994934082  segundos

#Para N=2^10 <hr>

---


Burbuja tardó: 			 20.698434829711914  	segundos

Insertion_Sort tardó: 		 0.058448076248168945  	segundos

Merge Sort tardó: 		 0.0059795379638671875  	segundos

El metodo sort de python tardó:  0.00018072128295898438  segundos

Merge Sort del Cormen tardó: 	 0.0052776336669921875  segundos

#Para N=2^11<hr>
Burbuja tardó: 			 170.51034808158875  	segundos

Insertion_Sort tardó: 		 0.25345635414123535  	segundos

Merge Sort tardó: 		 0.014197111129760742  	segundos

El metodo sort de python tardó:  0.0003025531768798828  segundos

Merge Sort del Cormen tardó: 	 0.011886358261108398  segundos